# Autobib

## EDA

We first perform an EDA of the dataset


In [2]:
import pandas as pd 

In [3]:
from data_manage import get_data, get_clean_average, preprocess

In [4]:
df = get_data('train')

In [5]:
df.head()

feature_site_D-8_installs  feature_site_D-8_d0_rev_value  \
0                        NaN                            NaN   
1                      805.0                       0.072786   
2                        NaN                            NaN   
3                        NaN                            NaN   
4                        1.0                       0.079453   

   feature_site_D-8_d8_rev_value  feature_site_D-7_installs  \
0                            NaN                        NaN   
1                       0.140335                      646.0   
2                            NaN                        NaN   
3                            NaN                        2.0   
4                       0.240620                        2.0   

   feature_site_D-7_d0_rev_value  feature_site_D-7_d7_rev_value  \
0                            NaN                            NaN   
1                       0.072781                       0.132105   
2                            NaN                            NaN   
3                       0.003011                       0.004328   
4                       0.048610                       0.059508   

   feature_site_D-6_installs  feature_site_D-6_d0_rev_value  \
0                        NaN                            NaN   
1                      687.0                       0.068517   
2                        NaN                            NaN   
3                        1.0                       0.001532   
4                        NaN                            NaN   

   feature_site_D-6_d6_rev_value  feature_site_D-5_installs  ...  \
0                            NaN                        NaN  ...   
1                       0.119165                      704.0  ...   
2                            NaN                        NaN  ...   
3                       0.001532                        NaN  ...   
4                            NaN                        NaN  ...   

   feature_app_D-3_d3_rev_value  feature_app_D-2_installs  \
0                      0.075404                   11594.0   
1                      0.052096                  200522.0   
2                      0.189784                    6782.0   
3                      0.045476                   40644.0   
4                      0.043391                  845087.0   

   feature_app_D-2_d0_rev_value  feature_app_D-2_d2_rev_value  \
0                      0.051442                      0.069153   
1                      0.028490                      0.046044   
2                      0.120070                      0.189299   
3                      0.024137                      0.032710   
4                      0.026870                      0.037267   

   feature_app_D-1_installs  feature_app_D-1_d0_rev_value  \
0                   11090.0                      0.052052   
1                  210970.0                      0.033250   
2                    6089.0                      0.125201   
3                   28002.0                      0.028119   
4                  720744.0                      0.026335   

   feature_app_D-1_d1_rev_value  feature_app_D0_installs  \
0                      0.063832                  10565.0   
1                      0.045672                 213128.0   
2                      0.168881                   5403.0   
3                      0.033978                  34619.0   
4                      0.032965                 765399.0   

   feature_app_D0_d0_rev_value  target_site  
0                     0.055874     0.026049  
1                     0.037260     0.243845  
2                     0.122818     0.552939  
3                     0.024765     0.023584  
4                     0.024237     0.529795  

[5 rows x 209 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Columns: 209 entries, feature_site_D-8_installs to target_site
dtypes: float64(209)
memory usage: 1.6 GB


In [7]:
print(df.mean(axis=0))

feature_site_D-8_installs            26.255127
feature_site_D-8_d0_rev_value         0.061138
feature_site_D-8_d8_rev_value         0.120903
feature_site_D-7_installs            25.955853
feature_site_D-7_d0_rev_value         0.061374
                                     ...      
feature_app_D-1_d0_rev_value          0.064787
feature_app_D-1_d1_rev_value          0.083085
feature_app_D0_installs          172255.377812
feature_app_D0_d0_rev_value           0.064103
target_site                           0.171790
Length: 209, dtype: float64


### EDA analysis
A first look at the data show that there are a lot of missing values. As we are dealing with numerical features the bulk of the preprocessing will consist in how we fill this missing values. It is tempting to fill them with 0, but there might be a better way to do so. 

For that purpose we dig further in the way the average data are distributed between the different columns. It would be interesting to see how the average value is distributed for each ganularity and each type ov feature: install features and revenue features 

#### Install features

In [8]:
df_install = get_clean_average(df, 'installs')
df_install

D-8_installs   D-7_installs   D-6_installs   D-5_installs  \
country_only   99468.148125   99513.548513   99772.388495  100207.097962   
country         6593.041653    6718.015306    6851.761743    7003.203284   
site              26.255127      25.955853      25.386049      24.535707   
campaign        1220.341027    1237.222376    1254.176167    1280.482410   
app_site         843.535940     836.855487     833.849105     828.697512   
app           150626.191571  152963.711743  155432.568412  158236.151515   
adgroup         1211.417223    1228.486862    1245.350780    1271.824379   
network         1340.148488    1357.753757    1376.781262    1403.788765   

               D-4_installs   D-3_installs   D-2_installs   D-1_installs  \
country_only  100573.668884  100720.058162  100738.952235  100791.056469   
country         7175.219712    7336.579654    7502.408824    7664.672497   
site              23.741092      23.199464      22.641877      21.903407   
campaign        1314.651172    1342.050342    1371.978999    1404.863469   
app_site         825.472142     819.025609     813.058810     803.824505   
app           161558.689150  164630.454675  167485.068976  170099.160788   
adgroup         1306.107124    1333.467379    1363.551591    1396.601720   
network         1439.339866    1467.755138    1499.823673    1535.772225   

                D0_installs  
country_only  100896.799504  
country         7838.711971  
site              21.287428  
campaign        1448.314063  
app_site         797.983251  
app           172255.377812  
adgroup         1440.262291  
network         1582.561222

We see that for a given granularity, the number of install is stable accross the different days considered on average

#### Revenue features

In [9]:
df_revenue = get_clean_average(df, 'value')
df_revenue

D-8_d0_rev_value  D-8_d8_rev_value  D-7_d0_rev_value  \
country_only          0.056106          0.111311          0.056211   
country               0.059789          0.118436          0.059945   
site                  0.061138          0.120903          0.061374   
campaign              0.060677          0.121034          0.060991   
app_site              0.063832          0.125183          0.064089   
app                   0.069600          0.137789          0.068131   
adgroup               0.060658          0.120978          0.060973   
network               0.060668          0.120966          0.060953   

              D-7_d7_rev_value  D-6_d0_rev_value  D-6_d6_rev_value  \
country_only          0.108271          0.056367          0.104638   
country               0.115331          0.060153          0.111566   
site                  0.118011          0.061999          0.114791   
campaign              0.117986          0.061345          0.114302   
app_site              0.122360          0.064160          0.118235   
app                   0.132093          0.067373          0.126508   
adgroup               0.117939          0.061325          0.114256   
network               0.117898          0.061257          0.114113   

              D-5_d0_rev_value  D-5_d5_rev_value  D-4_d0_rev_value  \
country_only          0.056415          0.100270          0.056357   
country               0.060275          0.107044          0.060179   
site                  0.062218          0.110370          0.062202   
campaign              0.061475          0.109635          0.061348   
app_site              0.064313          0.113702          0.064299   
app                   0.067185          0.120776          0.066872   
adgroup               0.061457          0.109595          0.061332   
network               0.061387          0.109506          0.061281   

              D-4_d4_rev_value  D-3_d0_rev_value  D-3_d3_rev_value  \
country_only          0.095296          0.056334          0.089495   
country               0.101765          0.060118          0.095494   
site                  0.105199          0.062165          0.098653   
campaign              0.104084          0.061245          0.097655   
app_site              0.108329          0.064185          0.101732   
app                   0.114208          0.066445          0.106369   
adgroup               0.104052          0.061227          0.097621   
network               0.103988          0.061198          0.097579   

              D-2_d0_rev_value  D-2_d2_rev_value  D-1_d0_rev_value  \
country_only          0.056308          0.082085          0.056321   
country               0.059996          0.087528          0.059960   
site                  0.062231          0.090724          0.062195   
campaign              0.061039          0.089326          0.060974   
app_site              0.064070          0.093325          0.064035   
app                   0.065634          0.096263          0.064787   
adgroup               0.061025          0.089305          0.060965   
network               0.060984          0.089276          0.060924   

              D-1_d1_rev_value  D0_d0_rev_value  
country_only          0.071928         0.056387  
country               0.076642         0.059961  
site                  0.079529         0.062261  
campaign              0.078122         0.060821  
app_site              0.081877         0.064118  
app                   0.083085         0.064103  
adgroup               0.078111         0.060810  
network               0.078080         0.060804

Interestingly, there seem to be some stability for each feature accross granularities in the Revenue case.

#### Pre-processing
Given these observation we'll handle null values differently for Revenue features and installs features.
Revenue features will have their null values replaced by the average of the corresponding column whereas install features will have their null values replaced by the average of the corresponding row

# Model Choice

Given that time is limited well compare different model performance with their default setting to choose our model

### Data
we reduce our training data to compare the models

In [10]:
df1 = df.sample(frac= 0.3)
df_clean = preprocess(df1)

In [11]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 300000 entries, 277000 to 721059
Columns: 209 entries, feature_site_D-8_installs to target_site
dtypes: float64(209)
memory usage: 480.7 MB


In [12]:
# lets check there is no null value left
df_clean.isna().values.any()

False

In [13]:
from sklearn.model_selection import train_test_split


In [14]:
X = df_clean
y = X.pop('target_site')
X_train, X_val, y_train, y_val = train_test_split(
        X, y, test_size=0.2)

In [15]:
# we perform some scaling on the data. It is not necessary for tree based models but we'll keep a common 
# groundwork for our analysis
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)


### Regression

regression will give us a benchmark. It's unlikely to be the most performant model, we should be able to do better with ensemble models or Neural Networks

In [16]:
from sklearn.linear_model import Ridge

# we impose a penalty using Ridge regression
reg = Ridge()
reg.fit(X_train, y_train)
r2 = reg.score(X_val, y_val)
print(f'Ridge regression r2: {r2}')

Ridge regression r2: 0.3521461225942428


### Random forests

In [17]:
#from sklearn.ensemble import RandomForestRegressor
#reg_rf = RandomForestRegressor(n_estimators=25)
#reg_rf.fit(X_train, y_train)
#r2 = reg_rf.score(X_val, y_val)
#print(f'Random forest r2: {r2}')

Random forest takes too much time we try Light GBM 

### Light GBM

In [20]:
from lightgbm import LGBMRegressor
reg_lg = LGBMRegressor(n_estimators=100)
reg_lg.fit(X_train, y_train)
r2 = reg_lg.score(X_val, y_val)
print(f'Light gbm r2: {r2}')

Light gbm r2: 0.36197570559213


Results so far don't look great we'll try some more preprocessing. Noticing that many columns in the revenu summary are pretty low, we could potentially drop them and reduce the dimensionality of our data. We could do some PCA also. 

For now as LightGBM looks slightly better we'll see if we can optimize it. We'll use all of our taining data.

### Hyperparameters tuning for LightGBM

In [28]:
df_clean = preprocess(df)
# we check the resulting dataframe has the proper shape
print(df_clean.shape)
# we check they are not null values left
print(df_clean.isna().values.any())

(1000000, 209)
False


In [29]:
# No need to scale the data as we're using a tree based model
X = df_clean
y = X.pop('target_site')
X_train, X_val, y_train, y_val = train_test_split(
        X, y, test_size=0.2)

In [30]:
reg_lg = LGBMRegressor(n_estimators=100)
reg_lg.fit(X_train, y_train)
r2 = reg_lg.score(X_val, y_val)
print(f'Light gbm r2: {r2}')

Light gbm r2: 0.3615388860545832


In [31]:
r2_training = reg_lg.score(X_train, y_train)
print(f'Light gbm r2 on training data: {r2_training}')

Light gbm r2 on training data: 0.38881527254117143


The performance on the training data is not too far to the one on the eval data, there doesn't seem to have an overfitting issue. We'll try to adjust max_depth (default -1) and n_estimators(defaut 100)

In [32]:
for n_estimator in [50, 100, 200]:
    for max_depth in [-1, 5, 10]:
        reg_lg = LGBMRegressor(n_estimators=n_estimator, max_depth=max_depth)
        reg_lg.fit(X_train, y_train)
        r2 = reg_lg.score(X_val, y_val)
        print(f'Light gbm r2 with n_estimator={n_estimator} and max_depth={max_depth}: {r2}') 

Light gbm r2 with n_estimator=50 and max_depth=-1: 0.359655617988591
Light gbm r2 with n_estimator=50 and max_depth=5: 0.35855956346878637
Light gbm r2 with n_estimator=50 and max_depth=10: 0.3599834789892621
Light gbm r2 with n_estimator=100 and max_depth=-1: 0.3615388860545832
Light gbm r2 with n_estimator=100 and max_depth=5: 0.3618232539083529
Light gbm r2 with n_estimator=100 and max_depth=10: 0.36216483021066637
Light gbm r2 with n_estimator=200 and max_depth=-1: 0.3633136739078455
Light gbm r2 with n_estimator=200 and max_depth=5: 0.3639265487871285
Light gbm r2 with n_estimator=200 and max_depth=10: 0.36387306213377624


In [33]:
for n_estimator in [200, 250, 300]:
    for max_depth in [2, 4, 6]:
        reg_lg = LGBMRegressor(n_estimators=n_estimator, max_depth=max_depth)
        reg_lg.fit(X_train, y_train)
        r2 = reg_lg.score(X_val, y_val)
        print(f'Light gbm r2 with n_estimator={n_estimator} and max_depth={max_depth}: {r2}') 

Light gbm r2 with n_estimator=200 and max_depth=2: 0.35406332119631323
Light gbm r2 with n_estimator=200 and max_depth=4: 0.3627851791470833
Light gbm r2 with n_estimator=200 and max_depth=6: 0.36476846247764405
Light gbm r2 with n_estimator=250 and max_depth=2: 0.355455269813172
Light gbm r2 with n_estimator=250 and max_depth=4: 0.363730356768292
Light gbm r2 with n_estimator=250 and max_depth=6: 0.3647833547752981
Light gbm r2 with n_estimator=300 and max_depth=2: 0.3564922983433856
Light gbm r2 with n_estimator=300 and max_depth=4: 0.36442834714203076
Light gbm r2 with n_estimator=300 and max_depth=6: 0.3645939263793777


The best result is obtained with n_estimator=200 and max_depth=6. That's the parameter we choose

### Final model calibration

To calibrate our final model we use all the available training data.

In [35]:
reg_lg = LGBMRegressor(n_estimators=200, max_depth=6)
reg_lg.fit(X, y)

LGBMRegressor(max_depth=6, n_estimators=200)

# Test results of the final model

In [38]:
df_test = get_data('eval')

In [39]:
df_test_clean = preprocess(df_test)
X = df_test_clean
y = X.pop('target_site')
print(X.shape)

(100000, 208)


In [40]:
r2_test = reg_lg.score(X, y)
print(f'Light gbm r2 test result: {r2_test}') 

Light gbm r2 test result: 0.34678620568966323


# Conclusion

The model selected is a LightGBM with n_estimators=200 and max_depth=6. It achieves a r2 of 0.35 on the test dataset which is similar to the performance reached during training.


# Further experiments

- Given the size of the dataset, it would have been interesting to explore Neural Networks. 
- It would have been interested to fine-tune other parameters for the LightGBM
- It could be interesting to see if by reducing the number of features we could get some improvement:
    - either by grouping some columns and average them
    - or by performing some PCA before training the model